lib/srv/desktop/windows_server.go

```diff
@@ -438,72 +438,86 @@ func (s *WindowsService) Serve(plainLis net.Listener) error {
+		if !ok {
+			return trace.ConnectionProblem(nil, "Got %T from TLS listener, expected *tls.Conn", conn)
 		}
 
-		go s.handleConnection(conn)
+		go s.handleConnection(proxyConn)
 	}
 }
 
 // handleConnection handles TLS connections from a Teleport proxy.
 // It authenticates and authorizes the connection, and then begins
 // translating the TDP messages from the proxy into native RDP.
-func (s *WindowsService) handleConnection(con net.Conn) {
-	defer con.Close()
+func (s *WindowsService) handleConnection(proxyConn *tls.Conn) {
+	defer proxyConn.Close()
 	log := s.cfg.Log
+	tdpConn := tdp.NewConn(proxyConn)
+
+	// Inline function to enforce that we are centralizing TDP Error sending in this function.
+	sendTdpError := func(message string) {
+		if err := tdpConn.OutputMessage(tdp.Error{Message: message}); err != nil {
+			s.cfg.Log.Errorf("Failed to send TDP error message %v: %v", tdp.Error{Message: message}, err)
+		}
+	}
 
 	// Check connection limits.
-	remoteAddr, _, err := net.SplitHostPort(con.RemoteAddr().String())
+	remoteAddr, _, err := net.SplitHostPort(proxyConn.RemoteAddr().String())
 	if err != nil {
-		log.WithError(err).Errorf("Could not parse client IP from %q", con.RemoteAddr().String())
+		log.WithError(err).Errorf("Could not parse client IP from %q", proxyConn.RemoteAddr().String())
+		sendTdpError("Internal error.")
 		return
 	}
 	log = log.WithField("client-ip", remoteAddr)
 	if err := s.cfg.ConnLimiter.AcquireConnection(remoteAddr); err != nil {
 		log.WithError(err).Warning("Connection limit exceeded, rejecting connection")
+		sendTdpError("Connection limit exceeded.")
 		return
 	}
 	defer s.cfg.ConnLimiter.ReleaseConnection(remoteAddr)
 
 	// Authenticate the client.
-	tlsConn, ok := con.(*tls.Conn)
-	if !ok {
-		log.Errorf("Got %T from TLS listener, expected *tls.Conn", con)
-		return
-	}
-	ctx, err := s.middleware.WrapContextWithUser(s.closeCtx, tlsConn)
+	ctx, err := s.middleware.WrapContextWithUser(s.closeCtx, proxyConn)
 	if err != nil {
 		log.WithError(err).Warning("mTLS authentication failed for incoming connection")
+		sendTdpError("Connection authentication failed.")
 		return
 	}
 	log.Debug("Authenticated Windows desktop connection")
 
 	authContext, err := s.cfg.Authorizer.Authorize(ctx)
 	if err != nil {
 		log.WithError(err).Warning("authorization failed for Windows desktop connection")
+		sendTdpError("Connection authorization failed.")
 		return
 	}
 
 	// Fetch the target desktop info. Name of the desktop is passed via SNI.
-	desktopName := strings.TrimSuffix(tlsConn.ConnectionState().ServerName, SNISuffix)
+	desktopName := strings.TrimSuffix(proxyConn.ConnectionState().ServerName, SNISuffix)
 	log = log.WithField("desktop-name", desktopName)
 
 	desktop, err := s.cfg.AccessPoint.GetWindowsDesktop(ctx, desktopName)
 	if err != nil {
-		log.WithError(err).Warning("Failed to fetch desktop by name")
+		log.WithError(err).Warningf("Failed to fetch desktop by name: %v", desktopName)
```

Comment:
No need to include `desktopName` here, because we've already added the `desktop-name` field to the logger above (`log.WithField("desktop-name", desktopName)`)